src/shogun/io/SGIO.h

```diff
@@ -535,39 +433,29 @@ class SGIO
 		/** whether syntax highlighting is enabled */
 		bool syntax_highlight;
 
-		/** log level */
-		EMessageType loglevel;
-		/** available log levels */
-		static const EMessageType levels[NUM_LOG_LEVELS];
-		/** message strings syntax highlighted*/
-		static const char* message_strings_highlighted[NUM_LOG_LEVELS];
-		/** message strings */
-		static const char* message_strings[NUM_LOG_LEVELS];
-
-	private:
-		RefCount* m_refcount;
-		/// function called to print normal messages
-		std::function<void(FILE*, const char*)> print_message;
-		/// function called to print warning messages
-		std::function<void(FILE*, const char*)> print_warning;
-		/// function called to print error messages
-		std::function<void(FILE*, const char*)> print_error;
+		class RedirectSink;
+		std::shared_ptr<RedirectSink> io_sink;
+		std::shared_ptr<spdlog::logger> io_logger;
+		std::shared_ptr<spdlog::details::thread_pool> thread_pool;
 };
 
 template <typename... Args>
-void SGIO::message(EMessageType prio, Args&&... args) const
+void SGIO::message(EMessageType prio, const SourceLocation& loc, const char* fmt, const Args&... args) const
 {
-	const auto& msg = format(prio, std::forward<Args>(args)...);
-	print(prio, msg);
+	if(should_log(prio))
+	{
+		fmt::memory_buffer msg;
+		fmt::format_to(msg, fmt, args...);
+		message_(prio, loc, fmt::string_view(msg.data(), msg.size()));
```

Comment:
do the formatting in the header, and the logging in the cpp file. slightly inefficient because `should_log` is called twice